collector/library/pod.js

```diff
@@ -66,72 +69,72 @@ class Pod {
 
         try {
             result = await fetch(this.targetUrl);
             response = await result.json();
         } catch (err) {
-            if (err.message.indexOf('ECONNREFUSED') !== -1) {
-                console.log('clearly no logs here: ', this.targetUrl);
-                console.log('shutting this pod off for now..');
-                this.state = 'disabled';
-                return;
-            }
+            //console.log('failed getting available batches', err);
+            return;
         }
 
         if ('status' in response && response.status === 'error') {
-            console.log('clearly no logs here: ', this.targetUrl);
-            console.log('shutting this pod off for now..');
-            this.state = 'disabled';
+            //console.log('found error when fetching availble batches', response);
             return;
         }
 
         const batches = sortBy(response, ['id']);
 
         for (let i = 0; i < batches.length; i++) {
-            let batch = batches[i];
-            let findResult = this.seenBatches.findIndex(b => b.id === batch.id);
-            if (findResult === -1) {
-                this.seenBatches.push(batch);
-                this.em.emit(BATCH, batch);
-            } else {
-                // Make sure that the batches that you've already seen have not changed in their length
-                // incase we have history for them on our file system
-
-                let bytesGotFromBatch = await this.getBytesSentFromThisParticularBatch({ batch });
-                // Incase we get disconnected for some reason (network failure or what not)
-                // Let's make sure that while we do still see this batch we make sure we have heard everything it has to say.
-                if (batch.batchSize > bytesGotFromBatch && this.unackedPackets.length === 0) {
-                    this.em.emit(BATCH, batch);
-                }
+            const batch = batches[i];
+            const bytesSentFromBatch = await this.getBytesSentFromThisParticularBatch({ batch });
+
+            if (bytesSentFromBatch > batch.batchSize) {
+                console.log('We collected more bytes than availble!', this.targetUrl, 'batch id:', batch.id, 'bytes handled:', bytesSentFromBatch, 'batch size:', batch.batchSize);
+                continue;
+            }
+
+            if (batch.batchSize > bytesSentFromBatch) {
+                console.log('downloading batch', batch);
+                // handle batch
+                await this.handleBatch(batch);
+
+                if (await this.getBytesSentFromThisParticularBatch({ batch }) < batch.batchSize) {
+                    console.log('could not download the entire batch', batch);
+                    return;
+                } // Avoid moving to next batch in case we could not download the entire batch
             }
         }
     }
 
+    stop() {
+        console.log(`Shutting down pod for endpoint ${this.targetUrl}`);
+        clearInterval(this.pid);
```

Comment:
this can lead to race conditions - its possible that right after this clearing we arrive at the finally below and get a new pid

Reply:
Good catch!
I'll make sure to set a property on the instance of the Pod. Checking it in the finally before I set the new pid.
That should do the trick I think.